addons/godot-next/nodes/behaviors.gd

```diff
@@ -0,0 +1,186 @@
+# author: xdgamestudios
+# license: MIT
+# description: Provides a behavior system API for the nodes. 
+# todo: Refactor code in general to clean and improve performance
+# usage:
+# - Creating:
+#     behaviors = Behaviors.new()
+# - Adding Behaviors:
+#     behaviors.add_behavior(MyBehavior) # Returns null if the user assigns an invalid Behavior class.
+#     Note:
+#     - Allows only one behavior of each type.
+#     - If behavior already exists pre-existing Behavior is returned.
+# - Checking Behaviors:
+#     behaviors.has_behavior(MyBehavior) # Returns true if behavior exists within the collection.
+# - Retrieving Behaviors:
+#     behaviors.get_behavior(MyBehavior) # Returns the existing instance of the behavior type.
+#     Note:
+#     - If behavior doesn't exist in the collection returns null.
+# - Removing Behaviors:
+#     behaviors.remove_behavior(MyBehavior) # Removes the Behavior from the collection.
+#     Note:
+#     - Returns true if successful. Else, returns false.
+tool
+extends Node
+class_name Behaviors
+
+##### PROPERTIES #####
+
+var _behaviors: ResourceSet = null
+
+var _callbacks: Dictionary = {
+	"_ready" : {},
+	"_process" : {},
+	"_physics_process" : {},
+	"_input" : {},
+	"_unhandled_input" : {},
+	"_unhandled_key_input" : {}
+}
+
+##### NOTIFICATIONS #####
+
+func _set(p_property: String, p_value) -> bool:
+	if _behaviors.can_contain(p_property):
+		#warning-ignore:return_value_discarded
+		_behaviors.set_element(p_property, p_value)
+		property_list_changed_notify()
+		return true
+	return false
+
+func _get(p_property: String):
+	if _behaviors.can_contain(p_property):
+		return _behaviors.get_element(p_property)
+	return null
+
+func _get_property_list() -> Array:
+	if not _behaviors:
+		_behaviors = ResourceSet.new()
+	_behaviors.setup("_behaviors", Behavior)
+	var list := []
+	list.append(PropertyInfo.new_dictionary("_behaviors", 0, "", PROPERTY_USAGE_STORAGE).to_dict())
+	list += _behaviors.get_collection_property_list()
+	return list
+
+func _ready() -> void:
+	if not Engine.editor_hint:
+		var behaviors = _behaviors.get_data()
+		for behavior in behaviors.values():
+			_initialize_behavior(behavior)
+		
+		for behavior in _callbacks["_ready"]:
+			behavior._ready()
+	_check_for_empty_callbacks()
+
+func _process(delta: float) -> void:
```

Comment:
We modified the `_callbacks` list so that it no longer needs to perform these `enabled` checks. Also, these enabled checks will fail since it is now `_enabled` and not `enabled`. 

Also, because each of these callbacks have only 1 parameter, let's just create a helper method to handle all of this re-used logic.

    func _handle_notification(p_name: String, p_param) -> void:
        for a_behavior in _callbacks[p_name]:
            a_behavior.call(p_name, p_param)
    
    func _process(delta: float) -> void:
        _handle_notification("_process", delta)

etc.